tests/new_api/test_device.py

```diff
@@ -249,6 +249,95 @@ def test_preprocess_correct_config_setup(self, config, expected_config):
+        else:
+            decomp = [op]
+
+        config = ExecutionConfig(gradient_method="best" if is_trainable else None)
+        program, _ = device.preprocess(config)
+        [new_tape], _ = program([tape])
+        expected_tape = qml.tape.QuantumScript([*decomp, qml.RX(1.23, wires=0)], tape.measurements)
+        assert qml.equal(new_tape, expected_tape)
+
+    @pytest.mark.parametrize(
+        "op, decomp_depth",
+        [
+            (qml.StatePrep([1 / np.sqrt(2), 1 / np.sqrt(2)], wires=0), 1),
+            (qml.StatePrep(np.array([1, 0]), wires=0), 1),
+            (qml.BasisState([1, 1], wires=[0, 1]), 1),
+            (qml.BasisState(qml.numpy.array([1, 1]), wires=[0, 1]), 1),
+            (qml.AmplitudeEmbedding([1 / np.sqrt(2), 1 / np.sqrt(2)], wires=0), 2),
+            (qml.MottonenStatePreparation([1 / np.sqrt(2), 1 / np.sqrt(2)], wires=0), 0),
+        ],
+    )
+    def test_preprocess_state_prep_middle_op_decomposition(self, op, decomp_depth):
+        """Test that state prep ops in the middle of a tape are always decomposed."""
+        tape = qml.tape.QuantumScript(
+            [qml.RX(1.23, wires=0), op, qml.CNOT([0, 1])], [qml.expval(qml.PauliZ(0))]
+        )
+        device = LightningDevice(wires=3)
+
+        for _ in range(decomp_depth):
+            op = op.decomposition()[0]
+        decomp = op.decomposition()
+
+        program, _ = device.preprocess()
+        [new_tape], _ = program([tape])
+        expected_tape = qml.tape.QuantumScript(
+            [qml.RX(1.23, wires=0), *decomp, qml.CNOT([0, 1])], tape.measurements
+        )
+        assert qml.equal(new_tape, expected_tape)
+
+    @pytest.mark.parametrize("wires", [5, 9, 10, 13])
+    def test_preprocess_qft_decomposition(self, wires):
+        """Test that qml.QFT is not decomposed for less than 10 wires."""
+        tape = qml.tape.QuantumScript(
+            [qml.QFT(wires=list(range(wires)))], [qml.expval(qml.PauliZ(0))]
+        )
+        dev = LightningDevice(wires=wires)
+
+        program, _ = dev.preprocess()
+        [new_tape], _ = program([tape])
+
+        if wires >= 10:
+            assert all(not isinstance(op, qml.QFT) for op in new_tape.operations)
+        else:
+            assert tape.operations == [qml.QFT(wires=list(range(wires)))]
+
+    @pytest.mark.parametrize("wires", [5, 10, 13, 15])
+    def test_preprocess_grover_operator_decomposition(self, wires):
+        """Test that qml.GroverOperator is not decomposed for less than 10 wires."""
+        tape = qml.tape.QuantumScript(
+            [qml.GroverOperator(wires=list(range(wires)))], [qml.expval(qml.PauliZ(0))]
+        )
+        dev = LightningDevice(wires=wires)
+
+        program, _ = dev.preprocess()
+        [new_tape], _ = program([tape])
+
+        if wires >= 13:
+            assert all(not isinstance(op, qml.GroverOperator) for op in new_tape.operations)
+        else:
+            assert tape.operations == [qml.GroverOperator(wires=list(range(wires)))]
+
```

Comment:
Could we add these tests to `test_templates.py`?

Reply:
I've left the first two tests in this file as `StatePrepBase` ops are not always templates.

Reply:
I've left the first two tests in this file as `StatePrepBase` ops are not always templates.